### PR TITLE
Make sure expired messages aren't served.

### PR DESCRIPTION
**Pull request by cpalone**
This PR changes psql's implementation of GetMessage() and Latest() such that expired messages are never served to users.

### CODE REVIEW

File: backend/psql/backend.go [hunk 1]
```diff
@@ -546,15 +546,35 @@ func (b *Backend) latest(ctx scope.Context, room *Room, n int, before snowflake.
...
+			args = append(args, before.String())
+		}
 	} else {
-		query = ("SELECT room, id, previous_edit_id, parent, posted, edited, deleted," +
-			" session_id, sender_id, sender_name, server_id, server_era, content, encryption_key_id" +
-			" FROM message WHERE room = $1 AND id < $3 AND deleted IS NULL ORDER BY id DESC LIMIT $2")
-		args = append(args, before.String())
+		threshold := time.Now().Add(time.Duration(-nDays) * 24 * time.Hour).Unix()
+		if before.IsZero() {
+			query = ("SELECT room, id, previous_edit_id, parent, posted, edited, deleted," +
```
Comment (logan) [inline backend/psql/backend.go]: This query is missing posted > $3

File: backend/psql/room.go [hunk 2]
```diff
@@ -38,18 +38,41 @@ type RoomBinding struct {
...
+
+	nDays, err := rb.DbMap.SelectInt("SELECT retention_days FROM room WHERE room = $1", rb.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	var query string
+	args := []interface{}{rb.Name, id.String()}
+
+	if nDays == 0 {
```
Comment (logan) [inline backend/psql/room.go]: Rather than duplicating the query, let's just fetch the message and check its timestamp on the server side.

**Comment by cpalone** [comment]
Ok, that's fixed now. :rat: was involved in my builds :/

**Comment by cpalone** [comment]
Fixed. Commenting for a notification since GithubBot was down for a code update.

File: backend/psql/backend.go [hunk 3]
```diff
@@ -546,15 +546,35 @@ func (b *Backend) latest(ctx scope.Context, room *Room, n int, before snowflake.
...
+				" session_id, sender_id, sender_name, server_id, server_era, content, encryption_key_id" +
+				" FROM message WHERE room = $1 AND id < $3 AND deleted IS NULL ORDER BY id DESC LIMIT $2")
+			args = append(args, before.String())
+		}
 	} else {
-		query = ("SELECT room, id, previous_edit_id, parent, posted, edited, deleted," +
-			" session_id, sender_id, sender_name, server_id, server_era, content, encryption_key_id" +
-			" FROM message WHERE room = $1 AND id < $3 AND deleted IS NULL ORDER BY id DESC LIMIT $2")
-		args = append(args, before.String())
+		threshold := time.Now().Add(time.Duration(-nDays) * 24 * time.Hour).Unix()
```
Comment (logan) [inline backend/psql/backend.go]: Need to lose the .Unix() here.

**Comment by logan** [comment]
Looks good, we're ready for a rebase+squash.

**Comment by logan** [comment]
Merged as 1b8f4b2.

**Pull request closed without merging**